neat_ml/model/train.py

```diff
@@ -0,0 +1,343 @@
+        - A dictionary of the best hyperparameters
+          found during the grid search.
+        - The predicted probabilities for the positive
+          class on the validation set.
+    """
+    spw = _scale_pos_weight(y_train)
+    print(f"[INFO] scale_pos_weight={spw:.3f}  |  train neg/pos={np.bincount(y_train)}")
+
+    param_names = list(_PARAM_GRID.keys())
+    param_values = list(_PARAM_GRID.values())
+    best_auc = -np.inf
+    best_params: Dict[str, int | float | None] = {}
+    best_model: Pipeline | None = None
+    val_proba_best = None
+
+    for param_comb in product(*param_values):
+        params = dict(zip(param_names, param_comb))
+        model = _build_pipeline(spw).set_params(**params)
+
+        model.fit(X_train, y_train)
+        val_proba = model.predict_proba(X_val)[:, 1]
+        auc = roc_auc_score(y_val, val_proba)
+        print(f"[GRID] params={params}  |  val ROC-AUC={auc:.4f}")
+
+        if auc > best_auc:
+            best_auc = auc
+            best_params = params
+            best_model = model
+            val_proba_best = val_proba
+
+    assert best_model is not None, "No model was fitted!"
+    assert val_proba_best is not None, "Validation probabilities were not calculated!"
+
+    pr_auc = average_precision_score(y_val, val_proba_best)
+    metrics = {"val_roc_auc": best_auc, "val_pr_auc": pr_auc}
+
+    print(f"[BEST] val ROC-AUC={best_auc:.4f}  | val PR-AUC={pr_auc:.4f}")
+    print(f"[BEST] hyper-parameters: {best_params}")
+
+    X_full = pd.concat([X_train, X_val], axis=0)
+    y_full = pd.concat([y_train, y_val], axis=0)
+    final_model = _build_pipeline(
+        _scale_pos_weight(y_full)
+        ).set_params(**best_params)
+    final_model.fit(X_full, y_full)
+
+    return final_model, metrics, best_params, val_proba_best
+
+
+def plot_roc(
+    y_true: np.ndarray | Sequence[int],
+    y_prob: np.ndarray | Sequence[float],
+    out_png: str,
+    label: str = "Validation",
+) -> None:
+    """
+    Generate and save a Receiver Operating Characteristic
+    (ROC) curve plot.
+
+    Calculates and plots the ROC curve and the Area Under
+    the Curve (AUC) score, then saves the figure to a PNG file.
+
+    Parameters
+    ----------
+    y_true : Sequence[int]
+        The true binary labels.
+    y_prob : Sequence[float]
+        The predicted probabilities for the positive class.
+    out_png : str
+        The file path where the output PNG image will be saved.
+    label : str, optional
+        The label for the ROC curve in the plot legend, by 
+        default "Validation".
+    """
+    fpr, tpr, _ = roc_curve(y_true, y_prob)
+    aucv = roc_auc_score(y_true, y_prob)
+    plt.figure(figsize=(4, 4))
+    plt.plot(fpr, tpr, lw=2, label=f"{label} AUC={aucv:.3f}")
+    plt.plot([0, 1], [0, 1], "--", lw=1, color="grey")
+    plt.xlabel("False-Positive Rate")
+    plt.ylabel("True-Positive Rate")
+    plt.legend()
+    plt.tight_layout()
+    plt.savefig(out_png, dpi=300)
```

Comment:
Let's avoid the "global"/MATLAB-like `plt` usage, instead favoring the object-oriented `fig`, `ax` approach..